src/api/informer.rs

```diff
@@ -180,27 +173,43 @@ impl<K> Informer<K> where
         let res = self.client.request::<ObjectList<Void>>(req).await?;
 
         let version = res.metadata.resourceVersion.unwrap_or_else(|| "0".into());
-        debug!("Got fresh resourceVersion={} for {}", version, self.resource.resource);
-        Ok( version )
+        debug!(
+            "Got fresh resourceVersion={} for {}",
+            version, self.resource.resource
+        );
+        Ok(version)
     }
 
     /// Watch helper
     async fn single_watch(&self) -> Result<(Vec<WatchEvent<K>>, String)> {
         let oldver = self.version();
         let req = self.resource.watch(&self.params, &oldver)?;
-        let events = self.client.request_events::<WatchEvent<K>>(req).await?;
+        let events = self
+            .client
+            .request_events::<WatchEvent<K>>(req)
+            .await?
+            .filter_map(|e| async move { e.ok() })
+            .collect::<Vec<_>>()
+            .await;
```

Comment:
This is presumably used straight by the Informer IIRC. Ideally, Informer should present a streaming event api as well, but you don't have to do that as part of the same PR, unless you're feeling ambitious.

Reply:
I'd be interested in giving that a shot, totally. This method here is used internally in `poll` only it seems, so would it be something along the lines of `single_watch` returning an `impl Stream<Item = WatchEvent<K>>` and then we handle that in `poll` to calculate the new version?

Reply:
Yeah, that's the right separation I think. We can handle the new version at each new steam item there, because each new item is the "last we've seen" so far :-)

Reply:
Okay I think I get it, I've updated the code as such I think. I also tried adding a `K: Unpin` bound to try to not use `boxed` but I get a crazy compiler error I don't understand, any ideas what this all means?

```
error[E0277]: the trait bound `std::future::GenFuture<[static generator@src/client/mod.rs:161:24: 173:14 resp:reqwest::async_impl::response::Response for<'r, 's, 't0> {&'r mut reqwest::async_impl::response::Response, reqwest::async_impl::response::Response, impl core::future::future::Future, impl core::future::future::Future, ()}]>: std::marker::Unpin` is not satisfied in `impl core::future::future::Future`
   --> src/api/informer.rs:139:52
    |
139 |                 while let Some(Ok(event)) = events.next().await {
    |                                                    ^^^^ within `impl core::future::future::Future`, the trait `std::marker::Unpin` is not implemented for `std::future::GenFuture<[static generator@src/client/mod.rs:161:24: 173:14 resp:reqwest::async_impl::response::Response for<'r, 's, 't0> {&'r mut reqwest::async_impl::response::Response, reqwest::async_impl::response::Response, impl core::future::future::Future, impl core::future::future::Future, ()}]>`
    |
    = help: the following implementations were found:
              <std::future::GenFuture<T> as std::marker::Unpin>
    = note: required because it appears within the type `impl core::future::future::Future`
    = note: required because of the requirements on the impl of `std::marker::Unpin` for `futures_util::stream::unfold::Unfold<reqwest::async_impl::response::Response, [closure@src/client/mod.rs:160:41: 174:10], impl core::future::future::Future>`
    = note: required because it appears within the type `impl futures_core::stream::Stream`
    = note: required because it appears within the type `impl futures_core::stream::Stream`

error[E0277]: the trait bound `std::future::GenFuture<[static generator@src/client/mod.rs:161:24: 173:14 resp:reqwest::async_impl::response::Response for<'r, 's, 't0> {&'r mut reqwest::async_impl::response::Response, reqwest::async_impl::response::Response, impl core::future::future::Future, impl core::future::future::Future, ()}]>: std::marker::Unpin` is not satisfied in `impl core::future::future::Future`
   --> src/api/informer.rs:139:45
    |
139 |                 while let Some(Ok(event)) = events.next().await {
    |                                             ^^^^^^^^^^^^^^^^^^^ within `impl core::future::future::Future`, the trait `std::marker::Unpin` is not implemented for `std::future::GenFuture<[static generator@src/client/mod.rs:161:24: 173:14 resp:reqwest::async_impl::response::Response for<'r, 's, 't0> {&'r mut reqwest::async_impl::response::Response, reqwest::async_impl::response::Response, impl core::future::future::Future, impl core::future::future::Future, ()}]>`
    |
    = help: the following implementations were found:
              <std::future::GenFuture<T> as std::marker::Unpin>
    = note: required because it appears within the type `impl core::future::future::Future`
    = note: required because of the requirements on the impl of `std::marker::Unpin` for `futures_util::stream::unfold::Unfold<reqwest::async_impl::response::Response, [closure@src/client/mod.rs:160:41: 174:10], impl core::future::future::Future>`
    = note: required because it appears within the type `impl futures_core::stream::Stream`
    = note: required because it appears within the type `impl futures_core::stream::Stream`
    = note: required because of the requirements on the impl of `core::future::future::Future` for `futures_util::stream::stream::next::Next<'_, impl futures_core::stream::Stream>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: could not compile `kube`.

To learn more, run the command again with --verbose.
```

Reply:
uhh, yeah, that's a tough one. :face_with_head_bandage:

It might be best to stick with the boxed solution for now.